### ISSUE TITLE
SLF4J_FORMAT_SHOULD_BE_CONST should trace caller, to stop warning if all callers use constant value

### ISSUE THREAD

**Issue by KengoTODA** [KengoTODA/findbugs-slf4j#35]
Currently this plugin marks following case as `SLF4J_FORMAT_SHOULD_BE_CONST`. refs: #32

```java
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class IssueDemonstrator {

    private void triggerWarning() {
        logMessage("This will trigger SLF4J_FORMAT_SHOULD_BE_CONST");
    }

    private void logMessage(String msg) {
        log.error(msg);
    }

}
```

But in this case, bytecode detector can trace method-call-tree, and verify that all callers uses constant as `msg`. This method works only for private methods.

**Comment by jeffjensen** [comment]
+1
Thank you for further attempts to resolve this issue!

**Comment by jeffjensen** [comment]
Below example gives false positive with 1.2.0.
1. Will your proposed change eliminate the false positive for this example?
2. With your proposed change, is the constant a problem here; need to convert constant to local variable?

```java
private static final String MSG = "{}: abcd={}"
    + ", really long text forcing line wraps for formatting={}" + ", y={}";

private void logMessage(final String a, final String b, final MyClass c) {
    final Y y = c.getX().getY();
    log.error(MSG, "literal", a, b, y);
}
```

**Comment by KengoTODA** [comment]
I cannot reproduce this false-positive by version `1.2.0`. You may check my unit tests in b55798f.

Could you try `javap -c` command to parse your class and share it with me?
In my environment (OSX Java `1.8.0_20`), `MSG` is treated as one long constant value even though we use `+` to concat literals.

```
$ javap -c test-case.target.classes.pkg.Issue35
Warning: Binary file test-case.target.classes.pkg.Issue35 contains pkg.Issue35
Compiled from "Issue35.java"
public class pkg.Issue35 {
  public pkg.Issue35();
    Code:
       0: aload_0
       :
       :
       :
      15: ldc           #8                  // String {}: abcd={}, really long text forcing line wraps for formatting={}, y={}
```

**Comment by KengoTODA** [comment]
@jeffjensen I have a question: Are you using SLF4J 1.6?
Travis CI reported that my test case fails only when it build with SLF4J 1.6.6. It might be the reason.

* https://travis-ci.org/KengoTODA/findbugs-slf4j/jobs/77889293

**Comment by jeffjensen** [comment]
I also created a simple class as above, stripped down from the class with the false positive, and the problem does not exist for me either.
Instead, I will email to you the source and .class file that encounters the false positive for your private testing instead of publicly posting it.  Thank you for looking into this!

**Comment by jeffjensen** [comment]
Not SLF4J 1.6; using SLF4J 1.7.10.

**Comment by jeffjensen** [comment]
In case it matters, we're using Java 1.7.0_71

**Comment by KengoTODA** [comment]
> Instead, I will email to you the source and .class file

Thanks, I have checked.

In my opinion, that is not false-positive. Here is simple code to reproduce:

```java
class Foo {
  private static final String MSG = "MSG FROM " + Foo.class.getSimpleName();

  void log() {
    logger.info(MSG);
  }
}
```

Java compiler cannot resolve value of this `MSG` at compile phase, and it should be resolved dynamically. Currently this plugin recommends following style (just use place-holder) to inserting dynamic values into log messages:

```java
class Foo {
  private static final String MSG = "MSG FROM {}";
  
  void log() {
    logger.info(MSG, Foo.class.getSimpleName());
  }
}
```

**Comment by jeffjensen** [comment]
You are so right - that's a runtime constant.  Sorry about my oversight and
thank you for reviewing.

**Comment by KengoTODA** [comment]
@jeffjensen Today I released version `1.2.1`, which solves this issue and #33. Please have a try. It will be published at Maven central several hours later.

**Comment by jeffjensen** [comment]
Thanks @KengoTODA, will test when it is in Central.
(FYI 1.2.1 not yet showing in Central: http://search.maven.org/#search|ga|1|g%3A%22jp.skypencil.findbugs.slf4j%22)

**Comment by jeffjensen** [comment]
1.2.1 works great with Java 7.  Thank you!

**Comment by KengoTODA** [comment]
:-) your welcome
